src/main/java/org/sonatype/nexus/repository/puppet/internal/util/PuppetDataAccess.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Sonatype Nexus (TM) Open Source Version
+ * Copyright (c) 2018-present Sonatype, Inc.
+ * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
+ *
+ * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
+ * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
+ *
+ * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
+ * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.repository.puppet.internal.util;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+
+import javax.annotation.Nullable;
+import javax.inject.Named;
+
+import org.sonatype.nexus.blobstore.api.Blob;
+import org.sonatype.nexus.common.collect.AttributesMap;
+import org.sonatype.nexus.common.hash.HashAlgorithm;
+import org.sonatype.nexus.repository.Repository;
+import org.sonatype.nexus.repository.storage.Asset;
+import org.sonatype.nexus.repository.storage.AssetBlob;
+import org.sonatype.nexus.repository.storage.Bucket;
+import org.sonatype.nexus.repository.storage.Component;
+import org.sonatype.nexus.repository.storage.MetadataNodeEntityAdapter;
+import org.sonatype.nexus.repository.storage.Query;
+import org.sonatype.nexus.repository.storage.StorageTx;
+import org.sonatype.nexus.repository.view.Content;
+import org.sonatype.nexus.repository.view.Payload;
+import org.sonatype.nexus.repository.view.payloads.BlobPayload;
+import org.sonatype.nexus.repository.view.payloads.StreamPayload;
+import org.sonatype.nexus.repository.view.payloads.StreamPayload.InputStreamSupplier;
+
+import com.google.common.base.Supplier;
+import com.google.common.collect.ImmutableList;
+
+import static java.util.Collections.singletonList;
+import static org.sonatype.nexus.repository.storage.ComponentEntityAdapter.P_VERSION;
+import static org.sonatype.nexus.repository.storage.MetadataNodeEntityAdapter.P_NAME;
+
+/**
+ * @since 0.0.1
+ */
+@Named
+public class PuppetDataAccess
```

Comment:
For hints on how to test a class like this see https://github.com/sonatype/nexus-internal/blob/9a1c6a600bbe1ebad0156b69cc9ac83bc550f4cb/private/plugins/nexus-repository-yum/src/test/java/org/sonatype/nexus/repository/yum/internal/database/YumQueryComponentTest.java#L64

Reply:
Taking a gander at that test class, you are testing a bunch of complicated methods (good!) and not the methods that we have in this class. I have personally always skipped over testing these methods because they are LARGELY the same from format to format, and they are almost akin to testing the StorageTxImpl stuff (which is already tested). Not a great excuse obviously, but they almost deserve being promoted into some other help library, with the ability to override if need be, from my perspective, as they've been pretty much the same in all the formats I've worked on.

Reply:
The challenge is this: I come along and want to add a new query to this class (or worse modify an existing method), first thing I'm going to want to do is add a test but wait.... no test class!?!?! wtf? Where's Jeffry!?!? 😄 We *get away with it* in our internal code because we wrap everything in integration tests (not that this is the answer) but we don't have that luxury here. These also go out to the community and the tests provide a level of documentation that Sonatype-outsiders just aren't going to get anywhere else.

I think we need to get out of the habit of skipping our unit tests because they haven't been done in the past, my gut feeling is that it is contributing to longer feedback loops (especially in the cases where we have to rely on an IT).

Reply:
Fair enough :)

Reply:
Some tests for this class here: 18ea5e3